pkg/ipam-controller/allocator/allocator.go

```diff
@@ -0,0 +1,393 @@
+				return retErr()
+			}
+		}
+	}
+	pa.allocations[allocData.Node] = r
+	return nil
+}
+
+func (pa *poolAllocator) checkAllocation(allocData nodeAllocationInfo) error {
+	if allocData.Subnet.String() != pa.cfg.Subnet.String() {
+		return fmt.Errorf("subnet mismatch")
+	}
+	if !allocData.Gateway.Equal(pa.cfg.Gateway) {
+		return fmt.Errorf("gateway mismatch")
+	}
+	if ip.Distance(allocData.StartIP, allocData.EndIP) != int64(pa.cfg.PerNodeBlockSize)-1 {
+		return fmt.Errorf("ip count mismatch")
+	}
+	return nil
+}
+
+func (pa *poolAllocator) getNodeAllocationInfo(node string, allocRange allocatedRange) nodeAllocationInfo {
+	return nodeAllocationInfo{allocatedRange: allocRange, Subnet: pa.cfg.Subnet, Gateway: pa.cfg.Gateway, Node: node}
+}
+
+// return slice with allocated ranges.
+// ranges are not overlap and are sorted, but there can be "holes" between ranges
+func (pa *poolAllocator) getAllocationsAsSlice() []allocatedRange {
+	allocatedRanges := make([]allocatedRange, 0, len(pa.allocations))
+	for _, a := range pa.allocations {
+		allocatedRanges = append(allocatedRanges, a)
+	}
+	sort.Slice(allocatedRanges, func(i, j int) bool {
+		return ip.Cmp(allocatedRanges[i].StartIP, allocatedRanges[j].StartIP) < 0
+	})
+	return allocatedRanges
+}
+
+// AllocationConfig contains configuration of the IP pool
+type AllocationConfig struct {
+	PoolName         string
+	Subnet           *net.IPNet
+	Gateway          net.IP
+	PerNodeBlockSize int
+}
+
+func (pc *AllocationConfig) Equal(other *AllocationConfig) bool {
+	return reflect.DeepEqual(pc, other)
+}
+
+// New create and initialize new allocator
+func New() *Allocator {
+	return &Allocator{allocators: map[string]*poolAllocator{}}
+}
+
+type Allocator struct {
+	lock       sync.Mutex
+	allocators map[string]*poolAllocator
+	configured bool
+}
+
+// IsConfigured returns true if allocator is configured
+func (a *Allocator) IsConfigured() bool {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	return a.configured
+}
+
+// Configure update allocator configuration
+func (a *Allocator) Configure(ctx context.Context, configs []AllocationConfig) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	a.configure(ctx, configs)
+}
+
+// ConfigureAndLoadAllocations configures allocator and load data from the node objects
+func (a *Allocator) ConfigureAndLoadAllocations(ctx context.Context, configs []AllocationConfig, nodes []corev1.Node) {
+	log := logr.FromContextOrDiscard(ctx)
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	a.configure(ctx, configs)
+	for i := range nodes {
+		node := nodes[i]
+		nodePoolMgr, err := pool.NewManagerImpl(&node)
+		if err != nil {
+			log.Info("skip loading data from the node", "reason", err.Error())
+			continue
+		}
+		// load allocators only for know pools (pools which are defined in the config)
+		for poolName, poolData := range a.allocators {
+			nodeIPPoolConfig := nodePoolMgr.GetPoolByName(poolName)
+			allocInfo, err := ipPoolConfigToNodeAllocationInfo(node.Name, nodeIPPoolConfig)
```

Comment:
all of these scenarios (when we do continue) here are because of corrupt node annotation right ?
is there any other case where we would skip adding the allocation to the allocator for the pool ?

when node controller processes the node it will then re-allocate a new block.
which is interesting to test how CNI handles it on delete.

Reply:
Yes, if controller detects annotation which is not valid anymore it will immediately replace it. If pool config is required for CNIDel, then CNI plugin should cache this info

Reply:
ack, issue created.